sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+using syclext = sycl::ext::oneapi::experimental;
+
+struct Foo {
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void foo() {
+  double d = 3.14;
+}
+
+virtual SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY() void bar() {}
+};
+
+sycl::queue q(/* device selector choosing a device *without* fp64 support */);
+assert(!q.get_device().has(sycl::aspect::fp64));
+
+auto *Storage = sycl::malloc_device<Foo>(1, q);
+
+q.single_task([=] {
+  // The kernel is not submitted with 'calls_indirectly' property and therefore
+  // it is not considered to be using any of virtual member functions of 'Foo'.
+  // This means that the object of 'Foo' can be successfully created by this
+  // kernel, regardless of whether a target device supports 'fp64' aspect which
+  // is used by 'Foo::foo'.
+  // No exceptions are expected to be thrown.
+  new (Storage) Foo;
+});
+----
+
+==== Interaction with `reqd_sub_group_size` attribute
+
+The `reqd_sub_group_size` attribute is a bit of a special case comparing to
+other optional kernel features, because it requires to compile a kernel in a
+certain way, which may require special handling for all functions which are
+called from it.
+
+When the same helper function is called from two or more kernels
+with different `reqd_sub_group_size` attribute, it may be required for the
+implementation to duplicate that helper function to create different versions
+of it tailored to different sub-group sizes. It can be done in a straightforward
+manner when operating on a static call graph.
+
+Virtual member functions are essentially called indirectly and pointers to them
+are initialized just once when an object of a polymorhpic class is being
+created. Therefore, to support calling such virtual member function from two or
+more kernels with different `reqd_sub_group_size`, each kernel may need to
+receive a different pointer to a different version of a virtual member function.
+
+To avoid possibly posing such multi-versioning requirements on implementations,
+virtual member functions can only be called from kernels with _primary_
+sub-group-size as defined by
+link:../proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc[
+sycl_ext_oneapi_named_sub_group_sizes] extension, or otherwise behavior is
+undefined.
+
+NOTE: for implementations that don't support
+`sycl_ext_oneapi_named_sub_group_sizes` extension, virtual member functions can
+only be called from kernels which *don't* have `reqd_sub_group_size` attribute
+set on them explicitly, or otherwise behavior is undefined.
+
+=== Kernel bundles and device images
+
+When an object of a polymorphic class is constructed, it stores a pointer to
+virtual table, which points to its virtual member functions. Addresses of those
+functions are accessible and valid only within a kernel bundle containing a
+kernel which used to construct an object.
+
+Performing calls to virtual member functions of an object constructed in a
+kernel from a different kernel bundle is an undefined behavior.
```

Comment:
This limitation seems surprising to me.  If this is really a limitation, what happens if the user does *not* use the kernel bundle APIs but compiles the program with `-fsycl-device-code-split=per_kernel`?  This will cause every kernel to be in its own bundle.  If the application creates the object in one kernel and then uses the object in another kernel, won't we have this same situation where an object constructed in one bundle is used in another?

Reply:
> My expectation is that it would just work, because implementation will merge all dependent kernels and virtual functions together based on the new properties. But I must admit, that it is not spelled out anywhere in the spec.

I think you are suggesting that `sycl-post-link` will always bundle together into the same device image a kernel that constructs a polymorphic object along with all the kernels that could call virtual methods on that polymorphic object.  I think this could work, but I think it will get tricky.  Consider this example:

```
using syclext = sycl::ext::oneapi::experimental;

struct set1;
struct set2;

struct Foo {
  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<set1>)
  void foo() {}

  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<set2>)
  void bar() {}
};

int main() {
  sycl::queue q;

  auto *Storage = sycl::malloc_device<Foo>(1, q);

  q.single_task([=] {  // Kernel A
    new (Storage) Foo;
  });

  if (q.get_device().has(sycl::aspect::fp64)) {
    auto props = syclext::properties{syclext::calls_indirectly<set1>};
    q.single_task(props, [=] {  // Kernel B
      double d = 3.14;
      Storage->foo();
    });
  }

  auto props = syclext::properties{syclext::calls_indirectly<set2>};
  q.single_task(props, [=] {  // Kernel C
    Storage->bar();
  });
}
```

In this example kernels A, B and C may all need to be bundled together in order to ensure that the allocation of object `Foo` is in the same device image as the calls on its virtual member functions.  Will you be able to determine this because you will see the construction of `Foo` in kernel A and calls to virtual member functions in kernels B and C?

You will also have to *conditionally* bundle kernel B into that device image only if the device supports the `fp64` feature.  This can be done via dynamic linking, but it's a different use of dynamic linking beyond what you have described in the design document.

If this is too complicated, we could shift some of the responsibility to the user, and require the user to decorate kernel A with a property indicating the "set" that it belongs to.  It would then be the user's responsibility to ensure that the same "set" is used for the kernel that allocates the object and also for the kernels that calls virtual functions on the object.

BTW, a lot of the complexity comes from the fact that we allow one kernel to allocate a polymorphic object (via placement new) and then make virtual function calls on that object from a different kernel.  Do we know that this is actually an important requirement?  If we restricted virtual functions to the case when the object is allocated in the same kernel as the calls, the design would be simpler.

Reply:
> I think you are suggesting that `sycl-post-link` will always bundle together into the same device image a kernel that constructs a polymorphic object along with all the kernels that could call virtual methods on that polymorphic object.

Not necessarliy. My plan was, in fact, that runtime will be responsible for bundling all necessary kernels and virtual functions together - unless it is already explicitly done by user through explicit use of `kernel_bundle` API.

> BTW, a lot of the complexity comes from the fact that we allow one kernel to allocate a polymorphic object (via placement new) and then make virtual function calls on that object from a different kernel. Do we know that this is actually an important requirement?

We have been provided with the following application as basic acceptance example: [Exawind/exw-virtuals-app](https://github.com/Exawind/exw-virtuals-app) and it constructs a polymorphic object and uses its virtual member functions in separate kernels.

Reply:
> Not necessarliy. My plan was, in fact, that runtime will be responsible for bundling all necessary kernels and virtual functions together - unless it is already explicitly done by user through explicit use of `kernel_bundle` API.

When you say "runtime will be responsible for bundling all necessary kernels and virtual functions together", I assume you mean that the runtime will construct a single Level Zero or OpenCL module that contains all these kernels.  Right?  How can the runtime do this, though, in the AOT case?  In the AOT case, the kernels are already bundled into device images that have been AOT-compiled to native code.  We do not have any way at runtime to combine these native code modules together into a single module.

> We have been provided with the following application as basic acceptance example: [Exawind/exw-virtuals-app](https://github.com/Exawind/exw-virtuals-app) and it constructs a polymorphic object and uses its virtual member functions in separate kernels.

I was thinking about this more, and I think there is a problem even in a simple case.  Consider this simple case:

```
using syclext = sycl::ext::oneapi::experimental;

struct Foo {
  virtual
  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(syclext::indirectly_callable<>)
  void bar() {}
};

int main() {
  sycl::queue q;

  auto *Storage = sycl::malloc_device<Foo>(1, q);
  q.single_task([=] {  // Kernel A
    new (Storage) Foo;
  });

  // Many things happen ...

  auto props = syclext::properties{syclext::calls_indirectly<>};
  q.single_task(props, [=] {  // Kernel B
    Storage->bar();
  });
}
```

Let's assume the user is JIT-compiling and that there is a single SPIR-V module that contains kernels A and B.  When kernel A is JIT compiled, we cache a copy of the native code module.  The `Storage` object ends up embedding the address of "bar" as part of the v-table.

In my example, kernel B runs a long time after kernel A.  By the time this happens, the native code module may have been evicted from the cache.  Therefore, when kernel B runs we may JIT compile the SPIR-V module again.  However, I assume that the address of "bar" might be different in this new module.  Thus, it will be a problem when kernel B tries to use the v-table of the object in `Storage`.

I can think of other cases that will cause kernel B to be re-JIT-compiled too.  For example, if kernel B contains a specialization constant.

Reply:
I'm less concerned about in-memory cache, because we literally had only one case when we had to evict it and that is SYCL CTS for specialization constants, which is quite heavy. In any case, I've added some notes about it and specialization constants handling in eb5ea4d3dfbaae21c8df1629a163673bd7e423b1. I'm not sure if the suggested way of dealing with specialization constants is good enough, so I've also recorded it as an issue in 3728caeadfb218e8dc4370ae56d1e909a18ce9e4

Back to overall design:

> When you say "runtime will be responsible for bundling all necessary kernels and virtual functions together", I assume you mean that the runtime will construct a single Level Zero or OpenCL module that contains all these kernels. Right?

Right

> How can the runtime do this, though, in the AOT case?

In AOT case we should know target capabilities and therefore all the bundling will happen at compile-time. I.e. every device image that uses virtual functions is expected to be self-contained already as it is now.

With JIT flow, we don't know target capabilities in advance and all the bundling will happen at runtime as described in the design doc.